### PR TITLE
[terra-form-select] Add clear option

### LINKED ISSUE #1736: Form-Select: Need to be able to deselect
The problem:
Currently with any given select field, the user can not clear out their selection.  The placeholder text is not included in the list.

screen shot 2018-07-27 at 11 08 01 am
 
The enhancement:
For form selects, there are use cases where the user needs to be able to deselect the previously chosen selection (ie - clearing a filter, etc).

### PR DESCRIPTION
**Pull request by benbcai**
### Summary
- Added new props `allowClear` to show an option to allow clearing the selected item.

### Additional Details
Closed #1736. 
When the `allowClear` prop is set, an option is added to the top of the menu to allow clearing the selected item and reset the select back to the unselected state. The placeholder passed in is used as the display text for the clear option. If the placeholder is not provided or is an empty string, the default placeholder ("- Select -") would be used as the clear option text. The `allowClear` prop will be ignored for the `Mulitple` or `Tag` select variant.

Still requires some UX input on what margins to use for the clear option when displayed in an Option Group and accessibility when selecting the clear option.

### Example Deployment Links:
https://terra-core-deployed-pr-2240.herokuapp.com/#/tests/terra-form-select/form-select/allow-clear

Thanks for contributing to Terra. 
@cerner/terra

### CODE REVIEW

**Comment by StephenEsser** [comment]
Do the UX callouts mention a selection state for the clear option? Is the clear option always supposed to be selected by default?

image

**Comment by benbcai** [comment]
The terra-opensource-theme doesn't have the clear option selected by default. The other themes do have it in the selected state by default.

Open Source Theme:
screen shot 2019-02-19 at 12 11 05 pm

Clinical Theme:
screen shot 2019-02-19 at 12 11 24 pm

@neilpfeiffer Should the default selection state of the clear option be themeable? Or should we just choose one of the above and not make it themeable?

**Comment by benbcai** [comment]
@neilpfeiffer I'm still in need of a couple of things from UX.
 - margin-top and margin-bottom for the clear option in OptGroup.
 - After making the clear option to be not selected by default, the color contrast between the color and the hover background no longer meets the WCAG 2 AA contrast ratio thresholds. The color used to be #70787c, which is the same as the place holder color. I need a color that is a bit darker. I'm just using #666F70 for the time being.

screen shot 2019-03-01 at 2 00 29 pm

File: packages/terra-form-select/src/Select.jsx [hunk 1]
```diff
@@ -197,11 +203,20 @@ class Select extends React.Component {
...
-      children, defaultValue, onChange, placeholder, value, ...otherProps
+      allowClear, children, defaultValue, onChange, placeholder, value, ...otherProps
     } = this.props;
 
     const defaultPlaceholder = intl.formatMessage({ id: 'Terra.form.select.defaultDisplay' });
     const selectPlaceholder = placeholder === undefined ? defaultPlaceholder : placeholder;
+    let clearOptionDisplay;
+
+    if (allowClear) {
+      if (selectPlaceholder.length === 0) {
```
Comment (tbiethman) [inline packages/terra-form-select/src/Select.jsx]: If they give us an empty string, do we really want to override that?

Comment (benbcai) [reply packages/terra-form-select/src/Select.jsx]: We use the placeholder as the display for the clear option. If they provide an empty placeholder, we don't want the display for the clear option to also be empty. We are not overriding their placeholder, we are just falling back to our default placeholder (" -Select- ") as the clear option display if the placeholder they provide is really empty.

File: packages/terra-form-select/src/_ClearOption.module.scss [hunk 2]
```diff
@@ -0,0 +1,6 @@
+:local {
+  .clear-option {
+    color: var(--terra-form-select-clear-option-color, #666f70);
+    font-style: var(--terra-form-select-clear-option-font-style);
+  }
```
Comment (neilpfeiffer) [inline packages/terra-form-select/src/_ClearOption.module.scss]: Add margin:
```suggestion
    margin: var(--terra-form-select-clear-option-margin, 0 0 0.17857rem 0);
  }
```

File: packages/terra-form-select/src/Select.jsx [hunk 3]
```diff
@@ -9,6 +9,11 @@ import Tag from './_Tag';
 import Util from './_SelectUtil';
 
 const propTypes = {
+  /**
+   * Whether a clear option is available to clear the selection.
```
Comment (neilpfeiffer) [inline packages/terra-form-select/src/Select.jsx]:
```suggestion
   * Whether a clear option is available to clear the selection, will use placeholder text if provided.
```

**Comment by neilpfeiffer** [comment]
Couple of docsite adjustments: 
1. lets move `Clear` example before `Controlled`
https://github.com/cerner/terra-core/blob/5b9edded6d4cb2ed4a2fdcd21453d2b72e6c32db/packages/terra-form-select/src/terra-dev-site/doc/form-select/Select.1.doc.jsx#L101

2. Let update these example titles, so the set is more understandable which are types/variants, and which are ala carte features in addition to the variants:
- [ ]  `title: 'Disabled',` -> 'Disabled (applies to all)',
- [ ]  `title: 'Invalid',` -> 'Invalid (applies to all)',
- [ ]  `title: 'OptGroup',` -> 'Ability to add Option Groups (applies to all)'
- [ ]  `title: 'Custom Max Height',` -> 'Ability to set a Maximum Height (applies to all)'
- [ ]  `title: 'Clear',` -> ' Ability to Clear Selection (applies to only \'default\', \'combobox\', and \'search\')',
- [ ]  `title: 'Controlled',` -> 'Create a Controlled Select (applies to all)'

**Comment by mmalaker** [comment]
Functional verification completed. The select example functions as expected with the clear option. +1 to UX changes.

File: packages/terra-form-select/src/_ClearOption.module.scss [hunk 4]
```diff
@@ -0,0 +1,6 @@
+:local {
+  .clear-option {
+    color: var(--terra-form-select-clear-option-color, #666f70);
```
[the remark this answers is not in the document]
Comment (benbcai) [reply packages/terra-form-select/src/_ClearOption.module.scss]: Updated: https://github.com/cerner/terra-core/pull/2240/commits/d059316429c0bc319635619bbbc456c09589e49b

File: packages/terra-form-select/src/terra-dev-site/doc/form-select/Select.1.doc.jsx [hunk 5]
```diff
@@ -95,6 +97,11 @@ const Index = () => (
       description: 'An example of implementing a controlled Select.',
       example: <ControlledExample />,
       source: ControlledExampleSrc,
+    }, {
+      title: 'Clear',
+      description: 'An example with an option to clear the selected item.',
```
[the remark this answers is not in the document]
Comment (benbcai) [reply packages/terra-form-select/src/terra-dev-site/doc/form-select/Select.1.doc.jsx]: Updated: https://github.com/cerner/terra-core/pull/2240/commits/d059316429c0bc319635619bbbc456c09589e49b

File: packages/terra-form-select/src/_ClearOption.module.scss (hunk 2 above)
Comment (benbcai) [reply packages/terra-form-select/src/_ClearOption.module.scss]: Updated: https://github.com/cerner/terra-core/pull/2240/commits/d059316429c0bc319635619bbbc456c09589e49b

File: packages/terra-form-select/src/Select.jsx (hunk 3 above)
Comment (benbcai) [reply packages/terra-form-select/src/Select.jsx]: Updated: https://github.com/cerner/terra-core/pull/2240/commits/d059316429c0bc319635619bbbc456c09589e49b

**Comment by benbcai** [comment]
@neilpfeiffer 
- The Clear example has been moved to be before the Controlled example.
- The title and description on those examples are updated.

**Comment by neilpfeiffer** [comment]
Tested: PR #2240 
 
**Reviewed:**
- Resolves adding ability to soft "clear" a select option from the single select variants in  `terra-form-select`

**Related:**
- will need to add new variables and values into theme files 

\
\
Verified by @neilpfeiffer, ready for release.
 
**Pass** |
-----------|
UX Review|

**Pull request merged**